### PR TITLE
implement functor,apply for StateT

### LINKED ISSUE #405: a type error in A Magical Monad
```
Section:
  Hello World

Issue Type:
  Bug

Folder/File(s):
  - Hello-World/Application-Structure/src/MTL/Implementing-a-Monad-Transformer/Magical-Monad
 
```

In [this](https://github.com/JordanMartinez/purescript-jordans-reference/blob/latestRelease/21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md) file, ` g state` is a monad can't match to `(Tuple  value state2)`

image

I guess should change to something like this:

image

---

It's a wonderful, helpful project. Thanks.👏

### PR DESCRIPTION
**Pull request by JunYang-tes**
Fixes: #405

### CODE REVIEW

File: 21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md [hunk 1]
```diff
@@ -79,19 +79,18 @@ instance functor :: (Monad monad) => Functor (StateT state monad) where
       -- todo
     )
 
--- We need to use that function, but it only takes an `a`
--- argument. So, we need to get that `a` by using `g`
--- Thus, we'll pass the returning StateT's state argument into `g`
```
Comment (JordanMartinez) [inline 21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md]: Could you leave this comment as is, but include yours immediately below it?

File: 21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md [hunk 2]
```diff
@@ -192,53 +208,7 @@ instance ms :: (Monad m) => Monad (StateT state monad) where
...
-       -> (a -> StateT state monad b)
-       -> StateT state monad b
-  bind (StateT g) f = StateT (\s1 ->
-      (g s1) >>= (\(Tuple value1 s2) ->
-        let (StateT h) = f value1 in h s2
-      )
-    )
-
-instance monad :: (Monad m) => Monad (StateT state monad)
-```
```
Comment (JordanMartinez) [inline 21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md]: Why was this entire section deleted?

File: 21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md [hunk 3]
```diff
@@ -117,14 +115,34 @@ instance functor :: (Monad monad) => Functor (StateT state monad) where
...
+-- Finally, we can using `do notation` in this case
+instance functor :: (Monad monad) => Functor (StateT state monad) where
+  map :: forall a b
+       . (a -> b)
+      -> StateT state monad a
+      -> StateT state monad b
+  map f (StateT g) = StateT \state ->
+    do
+      (Tuple value state2) <- g state
+      pure $ Tuple (f value) state2
```
Comment (JordanMartinez) [inline 21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md]: I think I'd prefer this to show up at the very end. My goal here is not to be "readable" per say as much as it is to break down how the changes occur step by step. The FAABM section at the end shows the more readable version.

File: 21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md [hunk 4]
```diff
@@ -117,14 +115,34 @@ instance functor :: (Monad monad) => Functor (StateT state monad) where
...
+Actuclly, We don't even need the `monad` in `(StateT state monad)` to be a monad. What we did above is just take out the tuple from this monad and **map** it to a nother Tuple and than put the new tuple back to the same monad. Why don't we just **map** over it ? let's do it:
+```purescript
+instance functor :: (Functor f) => Functor (StateT state f) where
+  map :: forall a b
+      . (a -> b)
+      -> StateT state f a
+      -> StateT state f b
+  map aTob (StateT g) = StateT \state ->
+    map (\(Tuple a state2) -> Tuple (aTob a) state2) (g state)
+```
```
Comment (JordanMartinez) [inline 21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md]: While this is a good insight on your part, it's not my goal in this document to reveal that. I'd prefer either to not explain this insight here, or to explain it after the FAABM section.

**Review by JordanMartinez — CHANGES_REQUESTED** [review]
Thanks for the PR!

I'd prefer not to show how to implement the instances via do notation until the very end of the file. Could you only implement the change that we initially discussed in #405 and then continue discussing other changes to this file in a separate issue?

File: 21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md (hunk 2 above)
Comment (JunYang-tes) [reply 21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md]: Sorry about this. In last comment, I demonstrated how to use bind/>>= to implement functor and apply type class separately in related section. so I supposed this section is not required. 
I already added it back.

File: 21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md (hunk 4 above)
Comment (JunYang-tes) [reply 21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md]: Yeah, I removed it.

File: 21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md [hunk 5]
```diff
@@ -82,17 +82,32 @@ instance functor :: (Monad monad) => Functor (StateT state monad) where
 -- We need to use that function, but it only takes an `a`
 -- argument. So, we need to get that `a` by using `g`
 -- Thus, we'll pass the returning StateT's state argument into `g`
+-- Then we go a `monad (Tuple a state)`
```
Comment (JordanMartinez) [inline 21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md]: "go" -> "get"

File: 21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md [hunk 6]
```diff
@@ -82,17 +82,32 @@ instance functor :: (Monad monad) => Functor (StateT state monad) where
...
       -> StateT state monad a
       -> StateT state monad b
   map f (StateT g) = StateT (\state ->
       let
-        (Tuple value state2) = g state
+        ma = g state
       in
         -- todo
     )
+-- So we can using `bind/>>=` to opeartor the Tuple under this monad
```
Comment (JordanMartinez) [inline 21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md]: "using" -> "use"
"to opeartor" -> "to expose"
"Tuple under this" -> "Tuple within this"

File: 21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md [hunk 7]
```diff
@@ -102,10 +117,11 @@ instance functor :: (Monad monad) => Functor (StateT state monad) where
       -> StateT state monad b
   map f (StateT g) = StateT (\state ->
       let
-        (Tuple value state2) = g state
-        b = f value
+        ma = g state
       in
-        -- todo
+        ma >>= (\(Tuple value state2) ->
+          let b = f value in --todo
```
Comment (JordanMartinez) [inline 21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md]: `let b = f value in --todo`
->
```
let
 b = f value
in
 -- todo
```

File: 21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md [hunk 8]
```diff
@@ -118,10 +134,11 @@ instance functor :: (Monad monad) => Functor (StateT state monad) where
       -> StateT state monad b
   map f (StateT g) = StateT (\state ->
       let
-        (Tuple value state2) = g state
-        b = f value
+        ma = g state
       in
-        pure (Tuple b state2)
+        ma >>= (\(Tuple value state2) ->
+          let b = f value in pure $ Tuple b state
```
Comment (JordanMartinez) [inline 21-Hello-World/05-Application-Structure/src/02-MTL/02-Implementing-a-Monad-Transformer/03-A-Magical-Monad.md]: `let b = f value in pure $ Tuple b state`
->
```
let
 b = f value
in
 pure (Tuple b state2)
```

**Review by JordanMartinez — APPROVED** [review]
Thanks for making these changes. The work looks good.

I did notice a few other things that could be cleaned up (see the rest of this review). If you'd like to do that, then let me know. Otherwise, I'm ok with merging this PR and making those changes myself. I want to be respectful of the time you've already put into this PR, especially when these clarifications are very small and somewhat nitpicky.

**Comment by JordanMartinez** [comment]
Thanks for the PR

**Pull request merged**